src/neon/stage1.rs

```diff
@@ -0,0 +1,531 @@
+unsafe fn check_ascii_neon(si: &SimdInput) -> bool {
+    let high_bit: uint8x16_t = vdupq_n_u8(0x80);
+    let t0: uint8x16_t = vorrq_u8(si.v0, si.v1);
+    let t1: uint8x16_t = vorrq_u8(si.v2, si.v3);
+    let t3: uint8x16_t = vorrq_u8(t0, t1);
+    let t4: uint8x16_t = vandq_u8(t3, high_bit);
+
+    let v64: uint64x2_t = vreinterpretq_u64_u8(t4);
+    let v32: uint32x2_t = vqmovn_u64(v64);
+    let result: uint64x1_t = vreinterpret_u64_u32(v32);
+
+    vget_lane_u64(result, 0) == 0
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn check_utf8(
+    input: &SimdInput,
+    state: &mut Utf8CheckingState,
+) {
+    if check_ascii_neon(input) {
+        // All bytes are ascii. Therefore the byte that was just before must be
+        // ascii too. We only check the byte that was just before simd_input. Nines
+        // are arbitrary values.
+        let verror: int8x16_t =
+            int8x16_t::new(9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 1);
+        state.has_error =
+            vreinterpretq_s8_u8(vorrq_u8(vcgtq_s8(state.previous.carried_continuations, verror),
+                                         vreinterpretq_u8_s8(state.has_error)));
+    } else {
+        // it is not ascii so we have to do heavy work
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v0),
+                                          &mut (state.previous), &mut (state.has_error));
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v1),
+                                          &mut (state.previous), &mut (state.has_error));
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v2),
+                                          &mut (state.previous), &mut (state.has_error));
+        state.previous = check_utf8_bytes(vreinterpretq_s8_u8(input.v3),
+                                          &mut (state.previous), &mut (state.has_error));
+    }
+}
+
+// a straightforward comparison of a mask against input
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn cmp_mask_against_input(input: &SimdInput, m: u8) -> u64 {
+    macro_rules! call {
+        ($imm8:expr) => {
+            vmovq_n_u8($imm8)
+        };
+    };
+    let mask: uint8x16_t = constify_imm8!(m, call);
+
+    let cmp_res_0: uint8x16_t = vceqq_u8(input.v0, mask);
+    let cmp_res_1: uint8x16_t = vceqq_u8(input.v1, mask);
+    let cmp_res_2: uint8x16_t = vceqq_u8(input.v2, mask);
+    let cmp_res_3: uint8x16_t = vceqq_u8(input.v3, mask);
+
+    neon_movemask_bulk(cmp_res_0, cmp_res_1, cmp_res_2, cmp_res_3)
+}
+
+// find all values less than or equal than the content of maxval (using unsigned arithmetic)
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn unsigned_lteq_against_input(input: &SimdInput, maxval: u8) -> u64 {
+    macro_rules! call {
+        ($imm8:expr) => {
+            vmovq_n_u8($imm8)
+        };
+    };
+    let mask: uint8x16_t = constify_imm8!(maxval, call);
```

Comment:
What's the storry of the `constify_imm8` here, are we missing an intrinsic? I have the feeling a macro like that should never be called outside of stdlib/intrinsic code.

would:
```
let mask: uint8x16_t = vmovq_n_u8(maxval);
```
 do the trick?